azurelinuxagent/common/osutil/default.py

```diff
@@ -268,43 +271,29 @@ def remove_legacy_firewall_rule(self, dst_ip):
             logger.info(
                 "Unable to remove legacy firewall rule, won't try removing it again. Error: {0}".format(ustr(error)))
 
-    def enable_firewall(self, dst_ip, uid):
+    def enable_firewall(self, dst_ip, uid, wait):
         # If a previous attempt failed, do not retry
         global _enable_firewall  # pylint: disable=W0603
         if not _enable_firewall:
             return False
 
         try:
-            wait = self.get_firewall_will_wait()
-
-            # If the DROP rule exists, make no changes
+            # check every iptable rule and delete others if any rule is missing
+            #   and append every iptable rule to the end of the chain.
             try:
-                drop_rule = get_firewall_drop_command(wait, AddFirewallRules.CHECK_COMMAND, dst_ip)
-                shellutil.run_command(drop_rule)
-                logger.verbose("Firewall appears established")
-                return True
+                if not AddFirewallRules.verify_iptables_rules_exist(wait, dst_ip, uid):
+                    self.remove_firewall(dst_ip, uid)
```

Comment:
It does return code 1 if rule doesn't exist which was handled in `_delete_rule` function. 

`Additionally, I just checked, the function remove_firewall returns a boolean indicating whether it was able to delete the rules or not. If we're unable to delete the rules, are you just planning to add new rules irrespective?`
That's good point, it will mess up the rules and might have duplicate rules. I will update as If delete unsuccessful, we don't add rules.

Reply:
> I will update as If delete unsuccessful, we don't add rules.

The con of this would be that the agent would not be able to access wireserver, and even worse, wireserver might open up to non-root until we try again (probably in the next 30 secs). What would be the con of adding the same rules multiple times? Would it break firewall rules?

Reply:
Maybe consider adding a retry for deleting rules and if we're unable to delete it then just add the rules irrespective?

Reply:
`What would be the con of adding the same rules multiple times? Would it break firewall rules?`

It won't' break. Looking at the cons side I think you are right, will add the rules irrespective

Reply:
Make sure to test it out (that firewall rules are not being affected) :)

Reply:
Actually, someone thought about this before and added retry on delete rule. Kudos to them

Reply:
I just noticed `remove_firewall` does its own call for `wait = self.get_firewall_will_wait()`. Maybe you can remove this one since we're anyways fetching the one in `enable_firewall`?